portfolio/src/main/java/com/google/sps/servlets/PokedexDataServlet.java

```diff
@@ -0,0 +1,49 @@
+package com.google.sps.servlets;
+
+import com.google.gson.Gson;
+import java.io.IOException;
+import java.util.Scanner;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+@WebServlet("/pokedex-data")
+public class PokedexDataServlet extends HttpServlet {
+  private static final int NUMBER_OF_GENERATIONS =
+      7; // 6 generations, Array Index directly  (1st Gen -> index of 1)
+
+  private int getColIndex(String header, String colName) {
+    String[] colNames = header.split(",");
+    for (int i = 0; i < colNames.length; i++) {
+      if (colName.equals(colNames[i])) {
+        return i;
+      }
+    }
+    System.error.println("ERROR: Column Name given not valid")
+    return -1;
+  }
+
+  private int[] getGenerationCount() {
+    int[] result = new int[NUMBER_OF_GENERATIONS];
+    Scanner scanner = new Scanner(getServletContext().getResourceAsStream("/WEB-INF/pokemon.csv"));
+    String header = scanner.nextLine();
+    int generationColIndex = getColIndex(header, "Generation");
+    while (scanner.hasNextLine()) {
+      String line = scanner.nextLine();
+      String[] cells = line.split(",");
+      result[Integer.valueOf(cells[generationColIndex])]++;
+    }
+    scanner.close();
+    return result;
+  }
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
```

Comment:
I think this class could be made more readable by having the public method listed first, then getGenerationCount(), which is the first method called by the public method, and then getColIndex().

I checked the Java style guide and the external one just says to use "some logic order" (https://engdoc.corp.google.com/eng/doc/devguide/java/style/external.md#s3.4.2-ordering-class-contents)

However, the internal guide links to this doc with a lot more info, which I'm not sure you have access to:
http://go/java-practices/class-structure

If you don't have access, I think for now, you can gain a lot of readability by listing public methods first, followed by private methods in the order they are called by public methods, which makes it easier to read through a class top to bottom

Reply:
one tiny thing I'll say is I prefer if a private method is called only by one public method, to have that private method just below the public method. This helps in really large classes to prevent jumping around so much to trace the logic.

There is no hard & fast rule, it is really a judgment call.